text/four-russians.tex

```diff
@@ -0,0 +1,64 @@
+
+\section{Przyjęte oznaczenia}
+
+$A_n$ - n-ty znak słowa $A$\\
+$A^{i,j}$ - słowo $A_i...A_j$\\
+$D_a$ - koszt usunięcia znaku a\\
+$I_a$ - koszt wstawienia znaku a\\
+$R_{a,b}$ - koszt zamiany znaku a na znak b\\
+$\delta$ - odległość edycyjna, minimum po sumie kosztów wszystkich możliwych sekwencji edycyjnych,
+$\delta_{i,j}$ - odległość edycyjna między słowami $A^{1,i}$ oraz $B^{1,j}$\\
+
+\section{Algorytm Wagnera i Fischera}
+
+Algorytm opisany w pracy opiera się na pomyśle z innego algorytmu, autorstwa Wagnera i Fischera. Algorytm ten wyznacza odległość edycyjną dynamicznie, w czasie proporcjonalnym do iloczynu długości słów. Tworzy w tym celu pomocniczą macierz o rozmiarze $(|A|+1) \times (|B|+1)$, w której komórka (i,j) odpowiada kosztowi $\delta_{i,j}$. Poniższe twierdzenia prezentują sposób obliczania bazy i kolejnych komórek:\\
+
+\noindent \textbf{Twierdzenie 1}\\\\
+$\delta_{0,0} = 0$ $\forall i,j : 1 \leq i \leq |A|, 1 \leq j \leq |B|$\\
+$\delta_{i,0} = \sum_{1 \leq r \leq i}D_A_r$ oraz
+$\delta_{0,j} = \sum_{1 \leq r \leq j}I_B_r$\\
+
+\noindent \textbf{Twierdzenie 2}\\\\
+$\delta_{i, j} = min(\delta_{i-1,j-1}+R_{A_i,B_j}, \delta_{i-1,j}+D_A_i, \delta_{i,j-1}+I_B_j)$
+
+\section{Przyspieszenie metodą czterach Rosjan}
+
+Technika "czterech Rosjan" pochodzi z algorytmu obliczania przechodniego domknięcia grafu skierowanego poprzez podział macierzy sąsiedztwa na podmacierze o niewielkiej ilości wierszy. Okazuje się, że podobny pomysł można wykorzystać do asymptotycznego przyspieszenia powyższego algorytmu, właśnie dzięki podziałowi obliczenia na dużo mniejszych obliczeń.\\
+
+\noindent Zauważmy, że podmacierz $(i,j,m$) macierzy edycyjnej z algorytmu Wagnera-Fischera - zaczynająca się w komórce $(i+1, j+1)$ i mająca rozmiar $m \times m$ - jest zdeterminowana tylko i wyłącznie poprzez koszt $\delta(i,j)$ oraz wektory inicjalne $\delta(i,j+1)...\delta(i,j+m)$ oraz $\delta(i+1,j)...\delta(i+m,j)$. Obliczając wartość w komórce $(i,j)$ odwołujemy się bowiem tylko do trzech sąsiednich komórek $(i-1,j-1), (i-1,j), (i, j-1)$. Możemy więc wcześniej dla każdej możliwej podmacierzy $(i,j,m)$ obliczyć i zapamiętać jej wektory finalne: $\delta(i+m,j+1)...\delta(i+m,j+m)$ oraz $\delta(i+1,j+m)...\delta(i+m,j+m)$.\\
+
+\noindent Aby tego dokonać, musimy być w stanie wyenumerować wszystkie możliwe macierze rozmiaru $m$. Zakładamy, że alfabet jest skończony, więc słów długości $m$ jest skończona ilość. Jednak generowanie wszystkich możliwych wektorów inicjalnych może zająć zbyt dużo czasu - wartości komórek mają tendencję do wzrostu wraz z rozszerzaniem się macierzy głównej. Kiedy jednak ograniczymy funkcję kosztu do przypisywania operacjom edycyjnym wyłącznie wielokrotności jakiejś liczby rzeczywistej, okaże się, że mamy tylko skończoną ilość różnic pomiędzy sąsiednimi komórkami w macierzy edycyjnej.\\\\
```

Comment:
wyenumerować -> wygenerować
słów długości $m$ jest skończona ilość -> jest $|\A|^m$
skończoną ilość różnic -> skończoną liczbę różnic ("ilość" is in principle for uncountable stuff)